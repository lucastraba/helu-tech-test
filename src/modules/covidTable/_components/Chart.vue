<template>
  <div class="chart mx-auto">
    <VueApexCharts
      v-if="data && !errorMsg"
      type="bar"
      :options="options"
      :series="series"
    />
    <div
      v-else-if="loading"
      class="text-center text-danger my-2"
    >
      <b-spinner class="align-middle" />
      <strong>Loading...</strong>
    </div>
    <div
      v-else
      class="text-danger"
    >
      {{ errorMsg }}
    </div>
  </div>
</template>

<script>
import VueApexCharts from 'vue-apexcharts';
import TotalsMixin from '@/modules/covidTable/_components/TotalsMixin';

export default {
  name: 'Chart',
  components: {
    VueApexCharts
  },
  mixins: [TotalsMixin],
  props: {
    data: {
      type: Array,
      required: false,
      default: () => []
    },
    errorMsg: {
      type: String,
      required: false,
      default: ''
    },
    loading: {
      type: Boolean,
      required: false,
      default: false
    },
    suggestion: {
      type: String,
      required: false,
      default: ''
    }
  },
  data () {
    return {
      options: {
        chart: {
          id: 'covid-cases'
        },
        xaxis: {
          categories: ['Cases', 'Deaths', 'Recoveries']
        },
        yaxis: {
          labels: {
            formatter: (val) => {
              return new Intl.NumberFormat().format(val);
            }
          }
        },
        colors: ['#17A2B8', '#DC3545', '#28A745'],
        title: {
          text: 'Worldwide',
          align: 'center'
        },
        plotOptions: {
          bar: {
            borderRadius: 10,
            offsetX: 10,
            dataLabels: {
              position: 'top'
            }
          }
        },
        dataLabels: {
          enabled: true,
          formatter: (val) => {
            return new Intl.NumberFormat().format(val);
          },
          offsetY: -20,
          style: {
            fontSize: '12px',
            colors: ['#000000']
          }
        }
      }
    };
  },
  computed: {
    series () {
      return [{
        name: 'Cases',
        data: [this.calculateTotals('cases', this.data)]
      }, {
        name: 'Deaths',
        data: [this.calculateTotals('deaths', this.data)]
      }, {
        name: 'Recoveries',
        data: [this.calculateTotals('recoveries', this.data)]
      }];
    }
  }
};
</script>
<style scoped lang="scss">
.chart {
  max-width: 65rem;
  margin-top: 10px;
}
</style>
