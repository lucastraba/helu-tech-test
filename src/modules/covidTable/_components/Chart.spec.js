import { createLocalVue, shallowMount } from '@vue/test-utils';
import Chart from './Chart';

describe('Chart Component', () => {
  const localVue = createLocalVue();
  const factory = ({ propsData } = {}) => shallowMount(Chart, {
    localVue,
    stubs: {
      VueApexCharts: true,
      bSpinner: true
    },
    propsData: {
      ...propsData
    }
  });
  describe('conditional rendering', () => {
    it('should render the table if data is provided and there are no errors', () => {
      // Arrange.
      const wrapper = factory();
      // Act.
      // Assert.
      expect(wrapper.find('vueapexcharts-stub').exists()).toBe(true);
      expect(wrapper.find('b-spinner-stub').exists()).toBe(false);
      expect(wrapper.find('.error').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render the loading spinner if loading', () => {
      // Arrange.
      const wrapper = factory({
        propsData: {
          data: null,
          loading: true
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('vueapexcharts-stub').exists()).toBe(false);
      expect(wrapper.find('bspinner-stub').exists()).toBe(true);
      expect(wrapper.find('.error').exists()).toBe(false);
      expect(wrapper.html()).toMatchSnapshot();
    });
    it('should render an error if there is one', () => {
      const wrapper = factory({
        propsData: {
          errorMsg: 'error'
        }
      });
      // Act.
      // Assert.
      expect(wrapper.find('vueapexcharts-stub').exists()).toBe(false);
      expect(wrapper.find('b-spinner-stub').exists()).toBe(false);
      expect(wrapper.find('.error').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });
  });
  describe('computed', () => {
    describe('series', () => {
      it('should return a correct total of each dataset', () => {
        // Arrange.
        const data = [{
          cases: 1,
          deaths: 2,
          recoveries: 3
        }, {
          cases: 1,
          deaths: 2,
          recoveries: 3
        }, {
          cases: 1,
          deaths: 2,
          recoveries: 3
        }];
        const expected = [{
          name: 'Cases',
          data: [3]
        }, {
          name: 'Deaths',
          data: [6]
        }, {
          name: 'Recoveries',
          data: [9]
        }];
        const wrapper = factory({
          propsData: { data }
        });
        // Act.
        // Assert.
        expect(wrapper.vm.series).toStrictEqual(expected);
      });
    });
  });
});
