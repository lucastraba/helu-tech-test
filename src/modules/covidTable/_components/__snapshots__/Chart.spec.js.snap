// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`C Component conditional rendering should render an error if there is one 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"error text-danger\\">
    error
  </div>
</div>"
`;

exports[`C Component conditional rendering should render the loading spinner if loading 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"text-center text-danger my-2\\">
    <bspinner-stub class=\\"align-middle\\"></bspinner-stub> <strong>Loading...</strong>
  </div>
</div>"
`;

exports[`C Component conditional rendering should render the table if data is provided and there are no errors 1`] = `
"<div class=\\"chart mx-auto\\">
  <vueapexcharts-stub options=\\"[object Object]\\" type=\\"bar\\" series=\\"[object Object],[object Object],[object Object]\\" width=\\"100%\\" height=\\"auto\\"></vueapexcharts-stub>
</div>"
`;

exports[`Cha Component conditional rendering should render an error if there is one 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"error text-danger\\">
    error
  </div>
</div>"
`;

exports[`Cha Component conditional rendering should render the loading spinner if loading 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"text-center text-danger my-2\\">
    <bspinner-stub class=\\"align-middle\\"></bspinner-stub> <strong>Loading...</strong>
  </div>
</div>"
`;

exports[`Cha Component conditional rendering should render the table if data is provided and there are no errors 1`] = `
"<div class=\\"chart mx-auto\\">
  <vueapexcharts-stub options=\\"[object Object]\\" type=\\"bar\\" series=\\"[object Object],[object Object],[object Object]\\" width=\\"100%\\" height=\\"auto\\"></vueapexcharts-stub>
</div>"
`;

exports[`Char Component conditional rendering should render an error if there is one 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"error text-danger\\">
    error
  </div>
</div>"
`;

exports[`Char Component conditional rendering should render the loading spinner if loading 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"text-center text-danger my-2\\">
    <bspinner-stub class=\\"align-middle\\"></bspinner-stub> <strong>Loading...</strong>
  </div>
</div>"
`;

exports[`Char Component conditional rendering should render the table if data is provided and there are no errors 1`] = `
"<div class=\\"chart mx-auto\\">
  <vueapexcharts-stub options=\\"[object Object]\\" type=\\"bar\\" series=\\"[object Object],[object Object],[object Object]\\" width=\\"100%\\" height=\\"auto\\"></vueapexcharts-stub>
</div>"
`;

exports[`Chart Component conditional rendering should render an error if there is one 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"error text-danger\\">
    error
  </div>
</div>"
`;

exports[`Chart Component conditional rendering should render the loading spinner if loading 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"text-center text-danger my-2\\">
    <bspinner-stub class=\\"align-middle\\"></bspinner-stub> <strong>Loading...</strong>
  </div>
</div>"
`;

exports[`Chart Component conditional rendering should render the table if data is provided and there are no errors 1`] = `
"<div class=\\"chart mx-auto\\">
  <vueapexcharts-stub options=\\"[object Object]\\" type=\\"bar\\" series=\\"[object Object],[object Object],[object Object]\\" width=\\"100%\\" height=\\"auto\\"></vueapexcharts-stub>
</div>"
`;

exports[`NavBar Component conditional rendering should render an error if there is one 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"error text-danger\\">
    error
  </div>
</div>"
`;

exports[`NavBar Component conditional rendering should render the loading spinner if loading 1`] = `
"<div class=\\"chart mx-auto\\">
  <div class=\\"text-center text-danger my-2\\">
    <bspinner-stub class=\\"align-middle\\"></bspinner-stub> <strong>Loading...</strong>
  </div>
</div>"
`;

exports[`NavBar Component conditional rendering should render the table if data is provided and there are no errors 1`] = `
"<div class=\\"chart mx-auto\\">
  <vueapexcharts-stub options=\\"[object Object]\\" type=\\"bar\\" series=\\"[object Object],[object Object],[object Object]\\" width=\\"100%\\" height=\\"auto\\"></vueapexcharts-stub>
</div>"
`;
